---
title: "enrichTF-Introduction"
author: "Shining Ma"
date: "`r Sys.Date()`"
output: 
    html_document:
        toc: true
        toc_float: true
        number_sections: true
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{An Introduction to enrichTF}
  %\VignetteEncoding{UTF-8}
  
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Quick Start


## Download and Installation


The package enrichTF is part of Bioconductor project starting from Bioc 3.8 built on R 3.5. To install the latest version of enrichTF, please check your current Bioconductor version and R version first. The latest version of R is recommended, and then you can download and install enrichTF and all its dependencies as follows:

```{r installpkg,eval=FALSE,message=FALSE,warning=FALSE}
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("enrichTF")
```

## Loading
Similar with other R packages, please load enrichTF each time before using the package.

```{r loading00,eval=FALSE,message=FALSE}
library(enrichTF)
```
# Introduction
As transcription factors (TFs) play a crucial role in regulating the transcription process through binding on the genome alone or in a combinatorial manner, TF enrichment analysis is an efficient and important procedure to locate the candidate functional TFs from a set of experimentally defined regulatory regions. While it is commonly accepted that structurally related TFs may have similar binding preference to sequences (i.e. motifs) and one TF may have multiple motifs, TF enrichment analysis is much more challenging than motif enrichment analysis. Here we present a R package for TF enrichment analysis which combine motif enrichment with the PECA model.

# Workflow

## Prepare inputs
The function of genBackground is to generate random background regions based on the input regions and to set the length of input sequence regions. We try to select background sequence regions that match the GC-content distribution of the input sequence regions. The default length and number of background regions are 1000 and 10,000, respectively. The length of sequence regions used for motif finding is important. Here, we set 1000 as default value, which means for each region, we get sequences from -500 to +500 relative from center.
For example,
```{r loading01,eval=FALSE,message=FALSE}
setGenome("hg19")
foregroundBedPath <- system.file(package = "enrichTF", "extdata","testForeGround.bed")
gen <- genBackground(inputForegroundBed = foregroundBedPath,genome = "hg19")
```
## Motifscan
The function of MotifsInRegions is to scan for motif occurrences using the prepared PWMs and obtain the promising candidate motifs in these regions.
For example,
```{r loading02,eval=FALSE,message=FALSE}
setGenome("hg19")
foregroundBedPath <- system.file(package = "enrichTF", "extdata","testForeGround.bed")
gen <- genBackground(inputForegroundBed = foregroundBedPath)
findMotif <- enrichFindMotifsInRegions(gen,motifRc="integrate")
```
## Connect regions with target genes

The function of regionConnectTargetGene is to connect foreground and background regions to their target genes, which are predicted from PECA model.

For example,
```{r loading03,eval=FALSE,message=FALSE}
setGenome("hg19")
foregroundBedPath <- system.file(package = "enrichTF", "extdata","testForeGround.bed")
gen <- genBackground(inputForegroundBed = foregroundBedPath)
enrichFindMotifsInRegions(gen,motifRc="integrate")
```
## TF enrichment test
The function of TFsEnrichInRegions is to test whether each TF is enriched on input regions.
For example,
```{r loading04,eval=FALSE,message=FALSE}
setGenome("hg19")
foregroundBedPath <- system.file(package = "enrichTF", "extdata","testForeGround.bed")
gen <- genBackground(inputForegroundBed = foregroundBedPath)
conTG <- enrichRegionConnectTargetGene(gen)
findMotif <- enrichFindMotifsInRegions(gen,motifRc="integrate")
result <- enrichTFsEnrichInRegions(gen,findMotif,conTG)
```
